src/main/java/com/vaadin/training/bugrap/ui/ReportView.java

```diff
@@ -0,0 +1,113 @@
+package com.vaadin.training.bugrap.ui;
+
+import org.vaadin.bugrap.domain.entities.Report;
+import org.vaadin.bugrap.domain.entities.Reporter;
+
+import com.vaadin.data.Binder;
+import com.vaadin.data.ValidationException;
+import com.vaadin.navigator.View;
+import com.vaadin.navigator.ViewChangeListener.ViewChangeEvent;
+import com.vaadin.training.bugrap.scope.UIScope;
+import com.vaadin.training.bugrap.util.CommentFormat;
+import com.vaadin.ui.Button.ClickListener;
+
+public class ReportView extends ReportViewDesign implements View {
+
+	private static final long serialVersionUID = 1790972443114757675L;
+
+	private Binder<Report> reportBinder;
+	private ClickListener updateButtonListener;
+	private ClickListener revertButtonListener;
+
+	@Override
+	public void enter(final ViewChangeEvent event) {
+		initialize();
+	}
+
+	public void initialize() {
+		final BugrapApplicationModel applicationModel = getApplicationModel();
+		if (applicationModel.isSingleReportSelected()) {
+			final Report report = applicationModel.getReport();
+
+			reportSummaryLabel.setValue(report.getSummary());
+
+			initComboBoxes(applicationModel);
+
+			initComments(report);
+
+			initReportBinder(report);
+
+			initUpdateButton();
+			initRevertButton();
+		}
+	}
+
+	private void initComboBoxes(final BugrapApplicationModel applicationModel) {
+		prioritySelect.clear();
+		prioritySelect.setItems(applicationModel.listPriorities());
+
+		bugSelect.clear();
+		bugSelect.setItems(applicationModel.listBugTypes());
+
+		statusSelect.clear();
+		statusSelect.setItems(applicationModel.listStatus());
+
+		assignedToSelect.clear();
+		assignedToSelect.setItems(applicationModel.listReporters());
+
+		versionSelect.clear();
+		versionSelect.setItems(applicationModel.listProjectVersions());
+	}
+
+	private void initComments(final Report report) {
+		final String reportDescriptionComment = CommentFormat.format(getName(report.getAuthor()), report.getReportedTimestamp(),
+				report.getDescription());
+		commentsLabel.setValue(reportDescriptionComment);
+	}
+
+	private void initRevertButton() {
+		if (revertButtonListener == null) {
```

Comment:
The if-check is there because otherwise listener would get called multiple times? This could be solved in couple of other ways as well.
First would be to explicitly clear all the listeners from button before adding new (something like button.getListeners(ClickListener.class).forEach(item -> button.removeClickListener((Button.ClickListener) item)) might do the trick) .
Second (and nicer) way would be to initialize the view in constructor (so it would be initialized only once) and then use initialize-method to bind the selected report to existing structure.
Both of the ways would free us from tracking the state of the listeners.
And the same goes to update-button listener

Reply:
Ok!